cosmos/ethermint
# eth_getBlockByHash and block hash mapping functionality

- Edit: Implements #40 get block by hash
- Sets up mapping storage and query for #40 #43 and #30 
- Fixes linting issues
- Enable start and end blocker functions for evm module in Ethermint app

I included getBlockByHash implementation in this PR because it includes a fix for the hash returned by get block by number, we don't have access to previous block consensus hash to create mapping.

To test:
1. Run the Ethermint node:
```
make install 
rm -rf ~/.emint*
emintd init moniker --chain-id 3
emintcli config chain-id 3
emintcli config output json
emintcli config indent true
emintcli config trust-node true
emintcli keys add austin
testpass
testpass
emintcli emintkeys add austineth
testpass
testpass
emintd add-genesis-account $(emintcli keys show austin -a) 1000photon,100000000stake
emintd add-genesis-account $(emintcli emintkeys show austineth -a) 100000000photon,100000000stake
emintd gentx --name austin
testpass
emintd collect-gentxs
emintd validate-genesis
emintd start --pruning=nothing --rpc.unsafe --log_level "main:info,state:info,mempool:info,*:error"

```

2. Start rest server with unlocked key to send tx to test functionality
```
emintcli rest-server --laddr "tcp://localhost:8545"
```

3. Retrieve block by height first (easiest way to get a block hash lol) and to make sure that isn't broken:
```
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByNumber","params":["0x2", false],"id":1}' -H "Content-Type: application/json" http://localhost:8545/rpc
## response: {"jsonrpc":"2.0","id":1,"result":{"difficulty":null,"extraData":null,"gasLimit":"0xffffffffffffffff","gasUsed":null,"hash":"0x95c21627b2c775e97bd416c1f99e8526429d3d4f5a865e8f8a4f57216f6c33bb","logsBloom":"","miner":"0x25605df97e5b25f58d4a19ad7c7a78898237b9f410a912aa493cb25b80548547","nonce":null,"number":"0x2","parentHash":"0x8eb967a0ebfa222aa13dd898bf4110ed97a41354777f0f77f401c09066fc7d6e","sha3Uncles":null,"size":"0x229","stateRoot":"0xb9bea5f2b08231adbbe2cbed459c6835eb7914ef072474b0eec46e6cc225af7c","timestamp":"0x5d8d1494","totalDifficulty":null,"transactions":[],"transactionsRoot":"0x","uncles":null}}
```

Use the hash from the previous block, in this case it was 0x95c21627b2c775e97bd416c1f99e8526429d3d4f5a865e8f8a4f57216f6c33bb:
```
curl -X POST --data '{"jsonrpc":"2.0","method":"eth_getBlockByHash","params":["0x95c21627b2c775e97bd416c1f99e8526429d3d4f5a865e8f8a4f57216f6c33bb", false],"id":1}' -H "Content-Type: application/json" http://localhost:8545/rpc
## response: {"jsonrpc":"2.0","id":1,"result":{"difficulty":null,"extraData":null,"gasLimit":"0xffffffffffffffff","gasUsed":null,"hash":"0x95c21627b2c775e97bd416c1f99e8526429d3d4f5a865e8f8a4f57216f6c33bb","logsBloom":"","miner":"0x25605df97e5b25f58d4a19ad7c7a78898237b9f410a912aa493cb25b80548547","nonce":null,"number":"0x2","parentHash":"0x8eb967a0ebfa222aa13dd898bf4110ed97a41354777f0f77f401c09066fc7d6e","sha3Uncles":null,"size":"0x229","stateRoot":"0xb9bea5f2b08231adbbe2cbed459c6835eb7914ef072474b0eec46e6cc225af7c","timestamp":"0x5d8d1494","totalDifficulty":null,"transactions":[],"transactionsRoot":"0x","uncles":null}}
```
